### PR TITLE
13.4 2fa v2 maior

### PR DESCRIPTION
**Pull request by herakros**
You need an email that you can access. You need to check the ability to enable and disable two-factor authentication, and two-step login.
How two-factor authentication activation works:
1) You log in to your account
2) You must confirm your mail
3) In the profile you press the 2fa button and get the corresponding window (if you do not have 2fa activated you will immediately be sent a code to the mail, which will need to enter to activate, if you have already activated, you will be disabled and you will go through a similar step)

After activating 2fa to log in you need:
1) By default, enter the data in the login page
2) If you have activated 2fa you will be redirected to the code entry page, which will be sent to you in the mail, if the code is entered correctly, you will be authorized

### CODE REVIEW

File: src/app/core/guards/auth.guard.ts [hunk 1]
```diff
@@ -10,14 +10,19 @@ export class AuthGuard implements CanActivate {
...
+
+    if(await this.authenticationService.isTwoFactorAuthentucation())
+    {
+      return true;
+    }
+
     if(await this.authenticationService.isAuthenticatedWithRefreshToken())
     {
       return true;
     }
```
Comment (YevheniiVovk) [inline src/app/core/guards/auth.guard.ts:18]: return await this.authenticationService.isAuthenticatedWithRefreshToken()

File: src/app/core/models/twoFactorDTO.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+export class TwoFactorDTO{
+    email?: string;
+    provider?: string;
+    token?: string;
```
Comment (YevheniiVovk) [inline src/app/core/models/twoFactorDTO.ts]: Make same as server model

File: src/app/core/models/userChange2fa.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export class UserChange2fa{
+    token?: string;
```
Comment (YevheniiVovk) [inline src/app/core/models/userChange2fa.ts]: Same as previous

File: src/app/core/models/userTwoStepCode.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export class UserTwoStepCode{
+    token?: string;
```
Comment (YevheniiVovk) [inline src/app/core/models/userTwoStepCode.ts]: Same as previous

File: src/app/core/services/authentication.service.ts [hunk 5]
```diff
@@ -35,45 +37,81 @@ export class AuthenticationService {
...
-        localStorage.setItem('user', JSON.stringify(this.currentUser));
+      if(tokens.is2StepVerificationRequired && tokens.provider){
+        this.router.navigate(["twoStepVerification"], {
+          queryParams: {
+            provider: tokens.provider,
+            email: user.Email
+          }
+        });
+
+        localStorage.setItem('isTwoFactor', String(tokens.is2StepVerificationRequired));
```
Comment (YevheniiVovk) [inline src/app/core/services/authentication.service.ts]: Should we store configuration of auth in local storage ?

File: src/app/core/services/authentication.service.ts [hunk 6]
```diff
@@ -35,45 +37,81 @@ export class AuthenticationService {
...
       }
+
+      this.setTokensInLocalStorage(tokens);
     }));
   }
 
+  public twoStepLogin(twoFactor: TwoFactorDTO): Observable<void> {
+
+    return this.http.post<AuthResponse>(this.twoStepLoginUrl, twoFactor).pipe(map((tokens: AuthResponse) => {
+
```
Comment (YevheniiVovk) [inline src/app/core/services/authentication.service.ts:69]: not needed empty line

File: src/app/core/services/authentication.service.ts [hunk 7]
```diff
@@ -35,45 +37,81 @@ export class AuthenticationService {
...
   public async isAuthenticated(): Promise<boolean> {
     const token: any  = localStorage.getItem('token');
     const refreshToken: any = localStorage.getItem('refreshToken');
 
-    return token&& refreshToken;
+    return token && refreshToken;
   }
+
+  public async isTwoFactorAuthentucation(): Promise<boolean>{
+    const isTwoFacotr: any = localStorage.getItem('isTwoFactor');
```
Comment (YevheniiVovk) [inline src/app/core/services/authentication.service.ts]: Configuration in local storage, are we sure ?

File: src/app/presentation/components/change-two-fa/change-two-fa.component.html [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+                        <mat-label>Code</mat-label>
+                        <input type="text" matInput placeholder="Your code" #Token formControlName="Token" required>
+                    </mat-form-field>
+                    <button mat-raised-button class="w-100 font-weight-600">Ok</button>
+                  </div>
+                  <p class="form-title">Check your email address</p>
+
+              </form>
+      </div>
+
```
Comment (YevheniiVovk) [inline src/app/presentation/components/change-two-fa/change-two-fa.component.html]: tabulation

File: src/app/presentation/components/change-two-fa/change-two-fa.component.html [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+      <div class="form-block">
+              <form class="login-form" [formGroup]="twoFactorForm" novalidate (ngSubmit)="submit()">
```
Comment (YevheniiVovk) [inline src/app/presentation/components/change-two-fa/change-two-fa.component.html]: Are we sure that we haven't any validation for code like length or format ?

File: src/app/presentation/components/two-setp-verification/two-setp-verification.component.css [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    flex-direction: row;
+    align-content: flex-start;
+    justify-content: space-between;
+    padding: 0px 5px 15px;
+}
+.forgot-pass a{
+    color: #685e9d;
+}
+.forgot-pass a:hover {
+    color: #272e44 !important;
```
Comment (YevheniiVovk) [inline src/app/presentation/components/two-setp-verification/two-setp-verification.component.css]: Could you please rewrite this style with out using !important

File: src/app/presentation/components/two-setp-verification/two-setp-verification.component.html [hunk 11]
```diff
@@ -0,0 +1,22 @@
+<div class="container">
+  <div class="form-section">
+      <div class="form-block">
+          <div class="w-100 mb-20">
+              <h1 class="registration_title">PROVIS</h1>
+              <p class="registration_text">Best Task Manager</p>
+          </div>
+
+              <form class="login-form" [formGroup]="twoStepForm" novalidate (ngSubmit)="submit()">
```
Comment (YevheniiVovk) [inline src/app/presentation/components/two-setp-verification/two-setp-verification.component.html]: Are we sure that we haven't any validation for code like length or format ?

File: src/app/presentation/components/two-setp-verification/two-setp-verification.component.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  ngOnInit() {
+    Swal.fire({
+      title: 'Check your email address ' + this.activeRoute.snapshot.queryParams['email'],
+      text: "You need to copy code and enter it in this page!",
+      icon: 'warning',
+      showCancelButton: false,
+      confirmButtonColor: '#3085d6',
+      confirmButtonText: 'Ok i understand!'
```
Comment (YevheniiVovk) [inline src/app/presentation/components/two-setp-verification/two-setp-verification.component.ts]: What reason of object duplication ? 
Can we go forward by DRY principle  ?

File: src/app/presentation/components/user-profile/user-profile.component.ts [hunk 13]
```diff
@@ -120,6 +107,63 @@ export class UserProfileComponent implements OnInit {
...
+      res => {
+        if(!res){
+          Swal.fire({
+            title: 'Check your email address ' + this.userProfile.email,
+            text: "To activate two-factor authentication, enter the code sent to your mail",
+            icon: 'warning',
+            showCancelButton: false,
+            confirmButtonColor: '#3085d6',
+            confirmButtonText: 'Ok i understand!'
+          });
```
Comment (YevheniiVovk) [inline src/app/presentation/components/user-profile/user-profile.component.ts]: What reason of object duplication ? 
Can we go forward by DRY principle  ?

File: src/app/core/guards/auth.guard.ts (hunk 1 above)
Comment (herakros) [reply src/app/core/guards/auth.guard.ts:18]: I did not work on this method and not understanding the logic, made changes, I deleted my code and left it as it was. For the correct operation of the guard should be left as is.

File: src/app/presentation/components/change-two-fa/change-two-fa.component.html [hunk 14]
```diff
@@ -0,0 +1,14 @@
+<div class="form-block">
+    <form class="login-form" [formGroup]="twoFactorForm" novalidate (ngSubmit)="submit()">
+      <div style="padding:30px">
```
Comment (YevheniiVovk) [inline src/app/presentation/components/change-two-fa/change-two-fa.component.html]: Move into styles

File: src/app/core/validators/confirmCodeValidator.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import { Validators } from "@angular/forms";
+export class ConfirmCodeValidator{
+    public static getRequiredValidator()
+    {
+        return Validators.required;
+    }
+
+    public static getConfirmCode()
+    {
+      return [ConfirmCodeValidator.getRequiredValidator(), Validators.pattern(`^[ 0-9]+$`), Validators.maxLength(6), Validators.minLength(6)];
```
Comment (YevheniiVovk) [inline src/app/core/validators/confirmCodeValidator.ts]: Line length

File: src/app/presentation/components/login/login.component.ts [hunk 16]
```diff
@@ -50,10 +50,8 @@ export class LoginComponent implements OnInit {
             title: 'Login',
             text: "Success",
             showConfirmButton: false,
-            timer: 1000
+            timer: 1500
```
Comment (YevheniiVovk) [inline src/app/presentation/components/login/login.component.ts]: Consistency 
If we displayed all messages 1000ms we shouldn't display it 1500 in one place

File: src/app/presentation/components/two-step-verification/two-step-verification.component.html [hunk 17]
```diff
@@ -0,0 +1,21 @@
+<div class="container">
+  <div class="form-section">
+      <div class="form-block">
+          <div class="w-100 mb-20">
+              <h1 class="registration_title">PROVIS</h1>
+              <p class="registration_text">Best Task Manager</p>
+          </div>
+          <form class="login-form" [formGroup]="twoStepForm" novalidate (ngSubmit)="submit()">
+                <div style="padding:30px">
```
Comment (YevheniiVovk) [inline src/app/presentation/components/two-step-verification/two-step-verification.component.html]: Move styles into style css

File: src/app/presentation/components/change-two-fa/change-two-fa.component.html [hunk 18]
```diff
@@ -0,0 +1,14 @@
+<div class="form-block">
+    <form class="login-form" [formGroup]="twoFactorForm" novalidate (ngSubmit)="submit()">
+      <div class="div-padding">
```
Comment (YevheniiVovk) [inline src/app/presentation/components/change-two-fa/change-two-fa.component.html]: let's name class div-padding-30

File: src/app/presentation/components/two-step-verification/two-step-verification.component.html [hunk 19]
```diff
@@ -0,0 +1,21 @@
+<div class="container">
+  <div class="form-section">
+      <div class="form-block">
+          <div class="w-100 mb-20">
+              <h1 class="registration_title">PROVIS</h1>
+              <p class="registration_text">Best Task Manager</p>
+          </div>
+          <form class="login-form" [formGroup]="twoStepForm" novalidate (ngSubmit)="submit()">
+                <div class="div-padding">
```
Comment (YevheniiVovk) [inline src/app/presentation/components/two-step-verification/two-step-verification.component.html]: Same here

**Review by YevheniiVovk — APPROVED** [review]

**Review by AndriyKovalets — APPROVED** [review]

**Pull request merged**